internal/controller/postgrescluster/snapshots.go

```diff
@@ -133,49 +148,286 @@ func (r *Reconciler) reconcileVolumeSnapshots(ctx context.Context,
+// after a successful backup.
+func (r *Reconciler) reconcileDedicatedSnapshotVolume(
+	ctx context.Context, cluster *v1beta1.PostgresCluster,
+	instances *observedInstances,
+	clusterVolumes []corev1.PersistentVolumeClaim,
+) (*corev1.PersistentVolumeClaim, error) {
+
+	// Get feature gate state.
+	volumeSnapshotsFeatureEnabled := feature.Enabled(ctx, feature.VolumeSnapshots)
+
+	// Set appropriate labels for dedicated snapshot volume
+	labelMap := map[string]string{
+		naming.LabelCluster: cluster.Name,
+		naming.LabelRole:    naming.RoleSnapshot,
+		naming.LabelData:    naming.DataPostgres,
+	}
+
+	// If volume already exists, use existing name. Otherwise, generate a name.
+	var pvc *corev1.PersistentVolumeClaim
+	existingPVCName, err := getPGPVCName(labelMap, clusterVolumes)
+	if err != nil {
+		return nil, errors.WithStack(err)
+	}
+	if existingPVCName != "" {
+		pvc = &corev1.PersistentVolumeClaim{ObjectMeta: metav1.ObjectMeta{
+			Namespace: cluster.GetNamespace(),
+			Name:      existingPVCName,
+		}}
+	} else {
+		pvc = &corev1.PersistentVolumeClaim{ObjectMeta: naming.ClusterDedicatedSnapshotVolume(cluster)}
+	}
+	pvc.SetGroupVersionKind(corev1.SchemeGroupVersion.WithKind("PersistentVolumeClaim"))
 
-	// Find primary instance
-	primaryInstance := &Instance{}
-	for _, instance := range instances.forCluster {
-		if isPrimary, known := instance.IsPrimary(); isPrimary && known {
-			primaryInstance = instance
+	// If snapshots are disabled, delete the PVC if it exists and return early.
+	// Check the client cache first using Get.
+	if cluster.Spec.Backups.Snapshots == nil || !volumeSnapshotsFeatureEnabled {
+		key := client.ObjectKeyFromObject(pvc)
+		err := errors.WithStack(r.Client.Get(ctx, key, pvc))
+		if err == nil {
+			err = errors.WithStack(r.deleteControlled(ctx, cluster, pvc))
 		}
+		return nil, client.IgnoreNotFound(err)
+	}
+
+	// If we've got this far, snapshots are enabled so we should create/update/get
+	// the dedicated snapshot volume
+	pvc, err = r.createDedicatedSnapshotVolume(ctx, cluster, labelMap, pvc)
+	if err != nil {
+		return pvc, err
+	}
+
+	// Determine if we need to run a restore job, based on the most recent backup
+	// and an annotation on the PVC.
+
+	// Find the most recently completed backup job.
+	backupJob, err := r.getLatestCompleteBackupJob(ctx, cluster)
+	if err != nil {
+		return pvc, err
+	}
+
+	// Return early if no complete backup job is found.
+	if backupJob == nil {
+		return pvc, nil
 	}
-	// Return error if primary instance not found
-	if primaryInstance.Name == "" {
-		return nil, errors.New("Could not find primary instance. Cannot create volume snapshot.")
+
+	// If pvc has been annotated with latest backup job's completion time, then restore was
+	// successful and there's nothing left to do.
+	backupJobCompletionTime := backupJob.Status.CompletionTime.String()
+	// TODO: change this to check whether backupJobCompletionTime is more recent than pvc annotation.
```

Comment:
Is the `==` good enough here?

Reply:
I believe so... Unless someone manually changes the annotation value on the PVC to an incorrect timestamp it will always either be equal to the `backupJobCompletionTime`, indicating that the PVC is up to date, or have an earlier timestamp (!= `backupJobCompletionTime`), indicating that it needs to be restored.

Reply:
On second thought, I think `==` isn't good enough. Example scenario: A user is triggering backups with the manual process and has already had a successful backup that was then restored and snapshotted. When they trigger a new manual backup, the old one is deleted. Before the new backup completes, the only existing completed backup is the initial `replica-create` backup we always do on cluster creation. The timestamp for that initial backup is grabbed and is different than the timestamp in the pvc annotation so a restore job is created before the new manual backup completes.